alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/actions/navigationstrategies/KnownDestinationReaching.kt

```diff
@@ -0,0 +1,148 @@
+ */
+open class KnownDestinationReaching<T, N : Euclidean2DConvexShape, E>(
+    action: EuclideanNavigationAction<T, N, E, ConvexPolygon, Euclidean2DPassage>,
+    /**
+     * Known destinations.
+     */
+    private val destinations: List<Euclidean2DPosition>
+) : RouteFollowing<T, N, E>(
+    action,
+    emptyList<Euclidean2DPosition>().takeIf { destinations.isEmpty() } ?: with(action) {
+        val currPos = environment.getPosition(pedestrian)
+        val (closestDest, distanceToClosestDest) = destinations
+            .asSequence()
+            .map { it to it.distanceTo(currPos) }
+            .minBy { it.second }
+            ?: throw IllegalArgumentException("internal error: destinations can't be empty at this point")
+        destinations.asSequence()
+            .sortedBy { it.distanceTo(currPos) }
+            .map { it to findKnownPath(it, pedestrian, environment) }
+            .filter { (_, path) ->
+                /*
+                 * A path leading to a destination is considered "valid" when:
+                 * - it's not empty
+                 * - the path's start is closer to currPos than the closest destination (this
+                 * because otherwise it's more convenient to just pursue the latter).
+                 */
+                path.isNotEmpty() && currPos.distanceTo(path.first().centroid) < distanceToClosestDest
+            }
+            .map { (destination, path) -> path.map { it.centroid } + destination }
+            .firstOrNull() ?: listOf(closestDest)
+    }
+) {
+
+    /**
+     * A [waypoint] is considered reached when inside [currentRoom] or in an adjacent room.
+     */
+    override fun isReached(waypoint: Euclidean2DPosition, currentRoom: ConvexPolygon): Boolean = with(action) {
+        super.isReached(waypoint, currentRoom) || doorsInSight().map { it.head }.any { it.contains(waypoint) }
+    }
+}
+
+/**
+ * Finds a known path to the specified [destination]. The path is obtained from the [pedestrian]'s cognitive
+ * map and consists of a list of landmarks, it is computed so as to minimize the distance between the first
+ * landmark and the pedestrian's position and the distance between the last landmark and the destination. In
+ * spite of this, the first landmark may be far from the pedestrian's position and the same applies to the
+ * last landmark and the destination. This means the pedestrian may not have information regarding any path
+ * leading from its current position to the first landmark and from the last one to the destination (such a
+ * path may not exist at all).
+ * If the cognitive map is poor the returned path may be poor. On average when the first landmark is farther
+ * from the pedestrian's position than the destination itself it's just more convenient to pursue the latter.
+ * If the cognitive map is empty the path will be empty as well.
+ */
+fun <T, P, N : ConvexGeometricShape<P, *>, M : ConvexGeometricShape<P, *>> findKnownPath(
```

Comment:
put this one and the followign either inside the `companion`, or (if access to members is useful to reduce parameterization) as a function of the class. Protect the visibility if it is not used elsewhere (`private`).
On a second thought, definitely inside the class, to reuse all the generics.

`findKnownPath` can be an extension method of P

Reply:
and maybe at this point could be something like `P.knownPathTo`

Reply:
I moved both functions in the class above. 
I would avoid making `findKnownPath` an extension method of `P`, its signature now is very clean and I prefer having `findKnownPathTo(destination)` than `destination.knownPathTo()`.

Reply:
okay